src/pygnssutils/gnssntripclient.py

```diff
@@ -486,83 +477,134 @@ def _read_thread(
-                    elif rc == "1":  # retrieved sourcetable
-                        stopevent.set()
-                        self._connected = False
-                        self._app_update_status(False)
-                    else:  # error message
-                        stopevent.set()
-                        self._connected = False
-                        self._app_update_status(False, (f"Error!: {rc}", "red"))
-        except (
-            socket.gaierror,
-            ConnectionRefusedError,
-            ConnectionAbortedError,
-            ConnectionResetError,
-            BrokenPipeError,
-            TimeoutError,
-            OverflowError,
-        ):
-            stopevent.set()
-            self._connected = False
-        except ssl.SSLCertVerificationError as err:
-            tip = (
-                f" - try using '{server[4:]}' rather than '{server}' for the NTRIP caster URL"
-                if "certificate is not valid for 'www." in err.strerror
-                else (
-                    f" - try adding the NTRIP caster URL SSL certificate to {findcacerts()}"
-                    if "unable to get local issuer certificate" in err.strerror
-                    else ""
                 )
-            )
-            print(f"SSL Certificate Verification Error{tip}\n{err}")
-            stopevent.set()
-            self._connected = False
-            self._app_update_status(False, (f"Error!: {err.strerror[0:32]}", "red"))
+                self._do_log(
+                    f"SSL Certificate Verification Error{tip}\n{err}", VERBOSITY_LOW
+                )
+                self._retrycount = self._retries
+                stopevent.set()
+                self._connected = False
+                self._app_update_status(False, (f"Error!: {err.strerror[0:60]}", "red"))
+
+            except (
+                BrokenPipeError,
+                ConnectionAbortedError,
+                ConnectionRefusedError,
+                ConnectionResetError,
+                OverflowError,
+                socket.gaierror,
+                ssl.SSLError,
+                TimeoutError,
+            ) as err:
+                errm = str(repr(err))
+                erra = f"Connection Error {errm.split('(', 1)[0]}"
+                errl = f"Error connecting to {server}:{port}/{mountpoint}: {errm}"
+                if self._retrycount == self._retries:
+                    stopevent.set()
+                    self._connected = False
+                else:
+                    self._retrycount += 1
+                    errr = (
+                        f". Retrying in {self._retryinterval * self._retrycount} secs "
+                        f"({self._retrycount}/{self._retries}) ..."
+                    )
+                    erra += errr
+                    errl += errr
+                self._do_log(errl, VERBOSITY_LOW)
+                self._app_update_status(False, (erra, "red"))
+
+            sleep(self._retryinterval * self._retrycount)

```

Comment:
Hi, I have been using this updated gnssntripclient for a while and the retry mechanism works perfectly. However, I notice one thing that the client does not exit immediately after reaching the maximum retry. It will sleep for an extra "self._retryinterval*self._retries" period before it exits.

I am seeing this line will be executed when "if self._retrycount == self._retries" in line 529 and the stop event is set. Perhaps we should return straight away as long as the client is hitting the max retry ?

Reply:
https://github.com/semuconsulting/pygnssutils/commit/7f6871d273b6803e5161e5013e8c394e629af099